File: soyeun/경주로건설.py
from collections import deque


def solution(board):
    n = len(board)
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    def bfs(x, y, cost, d):
        graph = [[0] * n for _ in range(n)]

        for j in range(n):
            for k in range(n):
                if board[j][k] == 1:
                    graph[j][k] = 1

        queue = deque()
        queue.append([x, y, cost, d])

        while queue:
            x, y, cost, idx = queue.popleft()

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if nx < 0 or ny < 0 or nx >= n or ny >= n:
                    continue
                if graph[nx][ny]==1:
                    continue
                if idx == i:
                    new_cost = cost + 100
                else:
                    new_cost = cost + 600

                if graph[nx][ny] == 0 or ((graph[nx][ny] != 0) and (graph[nx][ny] > new_cost)):
                    graph[nx][ny] = new_cost
                    queue.append([nx, ny, new_cost, i])
                else:
                    continue
        return graph[n - 1][n - 1]

    return min(bfs(0, 0, 0, 1), bfs(0, 0, 0, 3))

File: miri/경주로건설.java
import java.util.*;
public class Solution {
    static int[][][] visited;
    static int N;
    static int[][] d = {{1,0},{-1,0},{0,-1},{0,1}};

    public int solution(int[][] board) {
        N = board.length;
        visited = new int[N][N][4];

        return bfs(board);
    }

    public class Node {
        int r, c, dir, cost;
        public Node(int r, int c, int dir, int cost) {
            this.r = r;
            this.c = c;
            this.dir = dir;
            this.cost = cost;
        }
    }

    public int bfs(int[][] board) {
        int x = 0, y = 0, direction = -1, cost = 0;
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(x,y,direction,cost));

        int min_cost = Integer.MAX_VALUE;

        while(!q.isEmpty()) {
            Node now = q.poll();

            if(now.r == N-1 && now.c == N-1) {
                min_cost = Math.min(min_cost, now.cost);
            }

            for(int dir = 0; dir < 4; dir++) {
                int dr = now.r + d[dir][0];
                int dc = now.c + d[dir][1];

                if(dr < 0 || dr >= N || dc < 0 || dc >= N || board[dr][dc] == 1){
                    continue;
                }

                int nextCost = now.cost;
                if(now.dir == -1 || now.dir == dir) {
                    nextCost += 100;
                } else {
                    nextCost += 600;
                }

                if(visited[dr][dc][dir] == 0 || board[dr][dc] >= nextCost) {
                    q.add(new Node(dr, dc, dir, nextCost));
                    visited[dr][dc][dir] = 1;
                    board[dr][dc] = nextCost;
                }
            }
        }
        return min_cost;
    }
}